WIP: Add download button in the ZoomWidget and select the ROI

Pierre-Assemat opened the pull request:
- [x] download view button
    - [x] Fix bug : Only right size of the image is download when zoom < 5
- [ ] ROI support
    - [x] download area button
    - [x] selection of an region of interest
    - [x] edit ROI dialog box
        - [x] edit magnification
        - [x] show Width and Height
        - [x] show Number of pixels 
        - [x] show Size of the file
        - [x] select Format (JPEG, PNG, TIF...)
    - [x] download edited ROI
    - [ ] debug 
        - [x] magnification impact the image output
        - [ ] ratio compression 
            - [ ] JPEG
            - [ ] PNG
            - [ ] TIFF
        - [x] crash of browser when submit with wrong value
        - [x] good value on Width, Height, and Size
    - [x] Re-organize the code following backbone structure
- [x] Give the same name for this two way of downloading the image
- [x] Add all the tests to assure the code coverage

Fixes : #352

jbeezley commented:
@Pierre-Assemat I added a commit that should get you started with the client side testing.  I left a couple of places for you to add some assertions.  Once that is done, we can take a look at where the coverage numbers are and think about where we can add more test cases.

Pierre-Assemat commented:
I write futher test, but I raise an error : 

![screenshot from 2017-08-11 14-54-43](https://user-images.githubusercontent.com/29961430/29230822-ac800ba8-7eb2-11e7-8fcb-4b43034b9bbb.png)

I don't really know why, do you have an idea ?

jbeezley commented:
I'm not sure.  What code did you add?

Pierre-Assemat commented:
How can I assure this PR to pass the tests ? 
`Test #29: web_client_HistomicsTK.annotations` is failed on travis but pass on local...

Do I need to update the branch with Master ?

jbeezley commented:
Sorry @Pierre-Assemat, I've been out.  I'm going to look into this error today.

Pierre-Assemat commented:
No it's ok, I asked David, and he explain me that it was caused by a PR #2273 and will be fix by the PR #2297 !

Pierre-Assemat commented:
I think there is still some work to do here, first the value of the size (bytes) isn't really close from the truth, then there are some visualisation bug on my instance, I don't know the origin, and if it's only in this PR, I prefere wait other review before merge it into master
What do you think @jbeezley @manthey ?

jcfr commented:
@Pierre-Assemat Well done, CI is green. Before merging, I suggest you squash together all commits into one. I can help you tomorrow with this.

And to capture the fact you worked with @jbeezley on this, you could add the following text at the end of the message:

```
This is the title of the commit

And some explaination

Co-authored-by: Jonathan Beezley <jonathan.beezley@kitware.com>
```

jbeezley commented on web_client/dialogs/editRegionOfInterest.js:
```diff
@@ -0,0 +1,207 @@
+            this.downloadDisable(true);
+        } else {
+            this.downloadDisable(false);
+        }
```
This can just be
```javascript
this.downloadDisable(Nbytes >= Math.pow(2, 30));
```

jbeezley commented on web_client/dialogs/editRegionOfInterest.js:
```diff
@@ -0,0 +1,207 @@
+     * Get the size of the file in the appropriate unity (Bytes, MB, GB...)
+     */
+    getConvertFileSize() {
+        var Nbytes = this.getFileSize();
```
Here an elsewhere, avoid making variables names starting with capital letters.  These are usually reserved for classes.

jbeezley commented on web_client/dialogs/editRegionOfInterest.js:
```diff
@@ -0,0 +1,207 @@
+            if (typeof element === typeof undefined) {
+                var msgDisable = $('<span></span>').text('Size > 1GB : Impossible Download ');
+                msgDisable.attr('id', 'msgDisable');
+                msgDisable.css({'color': 'red', 'margin-right': '120px'});
```
Move custom css to a stylesheet.  For the color, you can just add the `g-validation-failed-message` class to the element to keep the style consistent.

jbeezley commented on web_client/dialogs/editRegionOfInterest.js:
```diff
@@ -0,0 +1,207 @@
+            }
+        } else if (bool === false) {
+            $('#download-submit').removeAttr('disabled');
+            if (typeof element !== typeof undefined) {
```
Normally you would want to use `_.isUndefined(element)` to check if a variable is undefined.  However in this case, I think the check you are doing is unnecessary.  It's usually better to allow a bug to throw an error, unless there is a specific reason to suppress it.

Pierre-Assemat replied:
I think I need this test, if the element don't exist and I don't want to disable the download
I enter in the 'else' and I check if the element : `this.$('#h-msgDisable')` exist before remove it
If I don't check the existance of the element, the remove will raise and error right ?

jbeezley replied:
No, `$('invalid selector').remove()` will not throw an error.

jbeezley commented on web_client/dialogs/editRegionOfInterest.js:
```diff
@@ -0,0 +1,207 @@
+                this._compressionRatio = 0.35;
+        }
+        this._magnification = parseFloat($('#h-element-mag').val());
+        var bounds = this.scaleBounds();
```
`var` definitions are always [hoisted](https://www.w3schools.com/js/js_hoisting.asp) to the top of the function.  It would be better to define the variables at the top or change all declarations to `let`/`const`.

jbeezley commented on web_client/panels/ZoomWidget.js:
```diff
@@ -166,6 +178,77 @@ var ZoomWidget = Panel.extend({
+            this.$('a.h-download-link#download-view-link').attr({
+                href: urlView
+            });
+        }
```
Setting an attribute is cheap.  I would just set it without trying to detect if it needs to be updated.

jbeezley commented on web_client/panels/ZoomWidget.js:
```diff
@@ -166,6 +178,77 @@ var ZoomWidget = Panel.extend({
+            contentDisposition: 'attachment'
+        });
+        var urlView = `/${apiRoot}/item/${imageId}/tiles/region?${params}`;
+        if (this._cancelSelection === true) {
```
`if (this._cancelSelection) {`

jbeezley commented on web_client/templates/dialogs/editRegionOfInterest.pug:
```diff
@@ -0,0 +1,39 @@
+          input#size-file.update-form.form-control(type='text', value=fileSize, disabled)
+        .g-validation-failed-message.hidden
+      .modal-footer
+        a#download-area-link.h-download-link.btn.btn-small.btn-primary
```
In Girder's ecosystem, we have a convention of prefixing class and id names with a letter.  In girder core, we use `g-`.  In HistomicsTK, it is `h-`.  Could you add `h-` to all of the custom names you defined?

Pierre-Assemat replied:
Thanks I didn't know !!

jbeezley commented on web_client/panels/ZoomWidget.js:
```diff
@@ -166,6 +178,77 @@ var ZoomWidget = Panel.extend({
+        if (this._cancelSelection === true) {
+            this.viewer.annotationLayer.mode(null);
+            this._cancelSelection = false;
+            $('.h-download-button-area').css('background-color', '');
```
Rather than set css on the element, just add or remove the `active` class on the button.

Also, here and elsewhere in this PR, you are referencing the global `$` object.  In most cases, this should be replaced with `this.$` to scope your selectors to just the current view.

manthey commented:
A few issues I've noticed:

When merged with the latest master, there is still a linting error and a test error.

I note that when you change the format control, it doesn't have any immediate affect -- the download still uses the previous format.  However, if you change the magnification, the format that becomes what was selected.

It behaves oddly on images that have no scale information (no magnification data).

Pierre-Assemat commented:
I found why the format wasn't update thank you @manthey !

When you say a linting error and test error, what do you mean ? A failure of a test ?

manthey commented:
The tests are failing on travis.  If you follow the latest travis PR log, you'll see
```
9: Test command: /home/travis/build/DigitalSlideArchive/HistomicsTK/build/girder/node_modules/.bin/eslint "--ignore-path" "/home/travis/build/DigitalSlideArchive/HistomicsTK/build/girder/.eslintignore" "--config" "/home/travis/build/DigitalSlideArchive/HistomicsTK/build/girder/.eslintrc.json" "/home/travis/build/DigitalSlideArchive/HistomicsTK/build/girder/plugins/HistomicsTK/web_client"
9: Test timeout computed to be: 1500
9: 
9: /home/travis/build/DigitalSlideArchive/HistomicsTK/build/girder/plugins/HistomicsTK/web_client/panels/ZoomWidget.js
9:   11:1  error  More than 1 blank line not allowed  no-multiple-empty-lines
9: 
9:  1 problem (1 error, 0 warnings)
9: 
26/31 Test  #9: eslint_HistomicsTK ...................................................***Failed    2.71 sec
```
You can also run these tests locally before you push a commit.

manthey commented on web_client/dialogs/editRegionOfInterest.js:
```diff
@@ -0,0 +1,194 @@
+                // JPEG is the default format
+                this._compressionRatio = 0.35;
+        }
+        this._magnification = parseFloat($('#h-element-mag').val());
```
The `const bounds = this.scaleBounds();` line above needs to come after the `_magnification` is set, otherwise the bounds  aren't correct.

manthey commented on web_client/panels/ZoomWidget.js:
```diff
@@ -166,6 +180,74 @@ var ZoomWidget = Panel.extend({
+    _downloadView(evt) {
+        var imageId = router.getQuery('image');
+        var bounds = router.getQuery('bounds');
+        var boundsTab = bounds.split(',');
```
There may not be any bounds in the route if an image was just loaded.

Pierre-Assemat replied:
I'm not sure to see what you're talking about, I tried to reproduce this in downloading an image that was just upload, this didn't cause any issue, and the bounds were set. What kind of changes could be more appreciate here ?

manthey replied:
I only saw it when I tried to merge the branch into master and get the tests to pass.  There was a merge conflict with some of the tests, and I may have not resolved them appropriately.

manthey replied:
We probably should be using `this.viewer.bounds()` rather then reparsing the router information.

Pierre-Assemat replied:
I'm resolving the conflict and fix the `annotations test`, I will see how I can use the `viewer.bounds` rather than using the router information. Thank you !

Pierre-Assemat commented:
Do you have any idea of what issue can cause this kind of display ?

![screenshot from 2017-10-16 13-32-09](https://user-images.githubusercontent.com/29961430/31627573-b8989034-b27b-11e7-9cc1-43ed9868f389.png)

manthey commented:
I'd guess that there is a javascript exception being thrown which stops some aspect of the renderer from running.

manthey commented:
Currently, we try to download a large region, the original image may need to access more than 0.5 gigapixels.  If so, PIL can throw an out-of-memory error, which results in a failure.  The replaces the HistomcsTK interface with the error message.  If we add a `target='_blank'` to the download link area, this at least appears in another window/tab (e.g., change the download link in pug to
```
a#h-download-area-link.h-download-link.btn.btn-small.btn-primary(target='_blank')
```
).  This isn't quite as nice looking as not opening a new tab, but guards against this error.

Pierre-Assemat commented:
This should happens only when we use the download button, but I have this kind of bad rendering each time I change the zoom value.

manthey commented:
@Pierre-Assemat Do you still have the problem with the zoom messing up the display?  Are you on the latest large_image and girder?  I haven't been able to reproduce the issue locally.

@jbeezley Would you care to take another look at this?

Pierre-Assemat commented:
@manthey Yes I have some weird behavior, I try to find the origin but I don't know.

First when I open the page the style is really weird but I don't see any error on the console log: 
![screenshot from 2017-11-07 15-18-34](https://user-images.githubusercontent.com/29961430/32515744-0b35ea8a-c3cf-11e7-9d6b-bfe5635583a6.png)

Then, each time the zoom is changed the render don't work well: 
![screenshot from 2017-11-07 15-21-39](https://user-images.githubusercontent.com/29961430/32515856-68d540aa-c3cf-11e7-883c-0d60f694533b.png)

I rebased on a master just after fetch it, I also updated girder, these errors has to come from one of the changes in the PR

jbeezley commented:
I'll give this branch a try.  If I can't reproduce it, we may need to do a hangout to debug what's going wrong in your development environment.

jbeezley commented:
This branch works for me as is.  I'll schedule a hangout with you to debug your development environment.

Pierre-Assemat commented:
Ok sounds great, thank you

manthey commented on plugin_tests/client/annotationSpec.js:
```diff
@@ -170,7 +169,91 @@ $(function () {
         });
+    });
+
+    describe('Download view and region of interest', function () {
```
Can you indent this entire describe block?  This will keep the outer describe block as a single block rather than divided into multiple describe blocks with the same description.

Pierre-Assemat replied:
Do you mean put this entire block into the discribe(Annotation tests) and remove the second discribe(Annotation tests) in order to keep only one block with this description ?

manthey replied:
Yes.  I'll just push the change.

manthey commented:
@jbeezley I've asked @Pierre-Assemat for one minor change in the layout of the tests, but otherwise I'm content to get this merged in if you are okay with it.  If we want further changes they can be in a different PR.
